### DISCUSSION TITLE
XCom values not being retrieved correctly between tasks in separate DAGs.

### DISCUSSION THREAD

**Discussion by adarshsaraogi** [apache/airflow#36438] [Q&A]
**Issue Summary:**
XCom values not being retrieved correctly between tasks in separate DAGs.

**Expected Behavior:**
I expect to retrieve the XCom value from Task B in DAG B after triggering it from Task A in DAG A.

**Actual Behavior:**
Despite successful execution of Task B and value pushing, Task A in DAG A is retrieving `None` when pulling the XCom value from Task B in DAG B.

**Configuration Details:**
- DAG A is triggering DAG B using `TriggerDagRunOperator`.
- Task B in DAG B is pushing an XCom value.
- Task A in DAG A is attempting to pull the XCom value from Task B.

**Steps to Reproduce:**
1. Trigger DAG B from Task A in DAG A.
2. Confirm successful execution of Task B in DAG B.
3. Attempt to pull the XCom value from Task B in DAG A.

**Code Snippets:**

*DAG A - Triggering DAG B:*
```python
# ... (relevant DAG A configurations)

trigger_dag_b = TriggerDagRunOperator(
    task_id='trigger_dag_b',
    trigger_dag_id='dag_b',
    conf={'key_to_pass': 'value_to_pass'},
    dag=dag_a,
)

# ... (rest of the DAG A configurations)

DAG B - Task B Pushing XCom Value:# ... (relevant DAG B configurations)

def process_and_pass_value(**kwargs):
    processed_value = "some_processed_value"
    kwargs['ti'].xcom_push(key='return_value', value=processed_value)

# ... (rest of the DAG B configurations)DAG A - Task A Pulling XCom Value:# ... (relevant DAG A configurations)

def retrieve_value(**kwargs):
    ti = kwargs['ti']
    retrieved_value = ti.xcom_pull(dag_id='dag_b', task_ids='process_and_pass_value', key='return_value')
    print(f"Retrieved Value from DAG B: {retrieved_value}")

# ...

**Comment by potiuk** [comment]
Read this https://medium.com/analytics-vidhya/airflow-xcom-pull-and-push-under-the-hood-multiple-value-from-different-dags-and-etc-2b9b3a942299 and likely there are plenty of other articles how you can read XCOM between dags. The key is executoon_date which is taken from the original task instance if you use the ti.xcom_pull

**Reply by adarshsaraogi** [reply]
@potiuk, I followed the article but my issue is when I trigger dag it should wait to get the value but I get none where my dag b is already having the value.

**Reply by potiuk** [reply]
Did you use the way to retrieve it described in the article? Your example shows different way of accessing xcom

**Reply by adarshsaraogi** [reply]
> Did you use the way to retrieve it described in the article? Your example shows different way of accessing xcom

@potiuk I have used this example where the execution date is implicitly passed by dag and I got correct results, but now it's not working as expected, however the code which I pasted was an example the actual code I am having is in company network I can share you the screenshot in some secure manner privately if you are okay, we have read the code where the execution date passed to xcom by context execution.

**Reply by potiuk** [reply]
No . Just look at the code and examples there to debug it  you need to pass the right execution date as in the link and many other examples.whwre people use XCOM class and execution date to retrieve the XCOM. Your issue is likely that you are not passing the right execution date to the call.

**Reply by adarshsaraogi** [reply]
@potiuk , even passing the write execution date still we are getting none value from xcom but in dag b task the xcom value is there.

**Reply by potiuk** [reply]
No idea - you need to debug it. It's easy, it converts to a database call, so you can trace it down to particular database query executed and see what mistake you've done.

Generally speaking - XCom is NOT ment to be cross-dag communication (as you can read in the links describing it). Currently the way to do cross-dag communication is to use datasets, but for those (for now) you should manually push and pull the data you want to use as inputs/outputs (in the future it will likely change).
